src/java/org/apache/cassandra/index/sai/memory/MultiBlockIndex.java

```diff
@@ -0,0 +1,165 @@
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.index.sai.memory;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import org.apache.cassandra.db.Clustering;
+import org.apache.cassandra.db.DecoratedKey;
+import org.apache.cassandra.db.PartitionPosition;
+import org.apache.cassandra.dht.AbstractBounds;
+import org.apache.cassandra.index.sai.IndexContext;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.index.sai.utils.PrimaryKeys;
+import org.apache.cassandra.index.sai.utils.RangeIterator;
+import org.apache.cassandra.index.sai.utils.RangeUnionIterator;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+
+/**
+ * Manages blocks of block indexes that are of size BLOCK_SIZE.  A block is filled to the block size,
+ * then a new block is used.
+ *
+ * At query time, blocks that fall within the key and value range are iterated on in merged primary key
+ * order until a match is found.
+ * @see BlockIndex
+ *
+ * Implements:
+ * @see MemoryIndex
+ */
+public class MultiBlockIndex extends MemoryIndex
+{
+    public static final int BLOCK_SIZE = 4000;
+    private final Object writeLock = new Object();
+    protected final ConcurrentLinkedQueue<BlockIndex> existingBlocks = new ConcurrentLinkedQueue<>();
+    protected BlockIndex current;
+
+    public MultiBlockIndex(IndexContext indexContext)
+    {
+        super(indexContext);
+    }
+
+    /**
+     * Add a key and value to the current block index.
+     * @param key Partition key
+     * @param clustering Clustering key
+     * @param value Index value
+     * @return Bytes memory used
+     */
+    @Override
+    public long add(DecoratedKey key, Clustering clustering, ByteBuffer value)
+    {
+        synchronized (writeLock)
+        {
+            if (current == null)
+                current = new BlockIndex(indexContext, this);
+
+            final long bytesUsed = current.add(key, clustering, value);
+
+            if (current.count() == BLOCK_SIZE)
+            {
+                existingBlocks.add(current);
+                current = null;
+            }
+            return bytesUsed;
```

Comment:
This comment does not make sense.

Reply:
Maybe I was unclear. Let's try again then.

* The javadoc on the parent (abstract) method does not exist. In particular, it does not document the semantics of the returned `long` value, and the meaning of it cannot be inferred from the method signature or the surrounding context.
* If the javadoc on the parent abstract method does exist, it technically allows the implementations can do anything. E.g. return -1, 0, Math.random() or just 2984935579. And such method would still comply to the contract. Yet obviously that would make no sense, so...
* ... I guess, the intent of the author of the abstract method was to return the number of memory bytes that have been consumed by the index so far. This is what `TrieMemoryIndex` and `BlockIndex` do.
* But the implementation of `MultiBlockIndex` seem to do a different thing. It returns the number of bytes used by the current block only (`currentIndex`), but ignores the bytes consumed by blocks stored in `existingBlocks`. So I asked if it is a bug, because it looks like at least an inconsistency.

Reply:
> It returns the number of bytes used by the current block only (currentIndex), but ignores the bytes consumed by blocks stored in existingBlocks.

This statement does not make sense.

Reply:
> This statement is not logical.

Please don't make vague statements like that with no justification. 
This is not bringing us closer to passing the review and closing the ticket.